File: ea-application/app-api/src/main/kotlin/com/nbe3/api/global/config/CursorArgumentResolver.java
package com.nbe3.api.global.config;

import com.nbe2.common.annotation.CursorDefault;
import com.nbe2.common.dto.Cursor;
import org.springframework.core.MethodParameter;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

public class CursorArgumentResolver implements HandlerMethodArgumentResolver {
    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(CursorDefault.class)
                && parameter.getParameterType().equals(Cursor.class);
    }

    @Override
    public Object resolveArgument(
            MethodParameter parameter,
            ModelAndViewContainer mavContainer,
            NativeWebRequest webRequest,
            WebDataBinderFactory binderFactory)
            throws Exception {
        CursorDefault cursorDefault = parameter.getParameterAnnotation(CursorDefault.class);

        String cursorParam = webRequest.getParameter("cursor");
        long cursor = (cursorParam == null) ? cursorDefault.cursor() : Long.parseLong(cursorParam);

        String sizeParam = webRequest.getParameter("size");
        int size = (sizeParam == null) ? cursorDefault.size() : Integer.parseInt(sizeParam);

        return new Cursor(cursor, size);
    }
}

File: ea-application/app-api/src/main/kotlin/com/nbe3/api/global/config/CursorArgumentResolver.kt
package com.nbe3.api.global.config

import com.nbe3.common.annotation.CursorDefault
import com.nbe3.common.dto.Cursor
import org.springframework.core.MethodParameter
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

class CursorArgumentResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.hasParameterAnnotation(CursorDefault::class.java)
                && parameter.parameterType == Cursor::class.java
    }

    @Throws(Exception::class)
    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?
    ): Any {
        val cursorDefault = parameter.getParameterAnnotation(CursorDefault::class.java)!!

        val cursorParam = webRequest.getParameter("cursor")
        val cursor = cursorParam?.toLongOrNull() ?: cursorDefault.cursor

        val sizeParam = webRequest.getParameter("size")
        val size = sizeParam?.toIntOrNull() ?: cursorDefault.size

        return Cursor(cursor, size)
    }
}
